### ISSUE TITLE
User authorization per instance of a type

### ISSUE THREAD

**Issue by ConstantinPatak** [MarimerLLC/cslaforum#119]
Hello everybody,

I have read in the CSLA books that the framework supports authorization per type (class) and per property of a class. Is it possible to give authorization per instance of a class. For example, lets say that you have a class BlogPost is there a way to assign read privileges to everybody and write privileges just to the creator of the BlogPost instance?

Thanks in advance,
Constantin

**Comment by jonnybee** [comment]
In my experience, I would rather override CanWriteProperty to add this as a custom check.

There is a limit of 1-one authorization rule per Type/Action and I find it much easier to have just one or two lines of code in CanWriteProperty that compare a Creator field with the current users name. Then use the standard authorization rules (and custom authz rules) on a per type basis.

**Comment by ConstantinPatak** [comment]
Thanks Jonny for your response.

**Comment by ajj3085** [comment]
I prefer to keep the logic in the auth rule if possible as it makes testing that logic easier.  What I've done is do the per type logic and set HasPermission appropriately, then check if the Target is not null and do processing which looks at the state of Target to set HasPermission using per instance logic.

**Comment by ConstantinPatak** [comment]
Thanks ajj3085 for your response. I am not sure if I completely understand the solution you purpose. I need to work on it. Thanks again.

**Comment by ajj3085** [comment]
Take this for example:
```
internal sealed class AuthorizeOwner : NonCachedAuthorizationRule {
	public AuthorizeOwner() : base(AuthorizationActions.EditObject) { }

	protected override void Execute(AuthorizationContext context) {
		if (context.Target == null) {
			context.HasPermission = true; // Users can edit orders in general
		} else {
			var order = (Order) context.Target;
			if (order.UserId == null) {
				throw new InvalidOperationException("Order does not have a UserId");
			}

			var currentUser = (ShoppingPricipal)ApplicationContext.User;
			context.HasPermission = currentUser.UserId == order.UserId;
		}
	}
}
```

Orders can be edited by users, but only the person that originally created the order can edit it, other user's can't.  The Order class would have this:

```
public static bool CanEditOrders() {
	return Csla.Rules.BusinessRules.HasPermission(AuthorizationActions.EditObject, typeof(Order));
}

public bool CanEdit() {
	return Csla.Rules.BusinessRules.HasPermission(AuthorizationActions.EditObject, this);
}
```

So it's possible for users to edit orders, but whether or not a particular order can be edited depends on the instance.

Hopefully that makes sense.  You could do other logic if `context.Target` is null to determine if the user can do anything at all with the type, and then if you have an instance you do more logic to see if they can edit the particular instance or not.

**Comment by jonnybee** [comment]
You might want to be more specific in your question. 

AuthorizationActions.EditObject is ONLY checked on root-level DataPortal.Update method. It is NOT checked for editing fields in the object. IE: A user may edit the fields as much as he/she likes but will get an exception on Save(). If the object is saved as a child object of another <root> object this permission is NOT checked by CSLA and Save will run OK. 

If you want the fields of the object to prevent writes you must either use the instance AuthorizationActions.WriteProperty and attach that authorization rule to ea ch property. And this is when it gets tricky if you have used other Authorization rules in your objects as there can be ONLY 1 - one Authorization rule per AuthorizationAction, Type, and property. 

And in these conditions I find it much easier to just have: 

        protected override bool CanWriteProperty(IPropertyInfo propertyInfo)
		{
            var currentUser = (ShoppingPricipal)ApplicationContext.User;
			// always readonly if not created by this user 
			if (currentUser.UserId != this.UserId) return false;

			// check regular authz rules 
			return base.CanWriteProperty(propertyInfo);
        }

This will prevent all writes in SetProperty and make CanWriteProperty return false (so you can also make the UI appear ReadOnly). 

You might want to use both techniques in these answers to get the behavior that you want.

**Comment by ajj3085** [comment]
Yes, Jonny brings up some good points, and you'll probably need to do a similar rule for fetch.  You'd have to call the CanGet() instance method yourself, likely in the DataPortal_Fetch, and after you've populated enough of the BO for the auth rule to be able to make a determination.  I have been assuming that you wouldn't let another user even fetch someone else's order, so per-property checks in that case might or might not be required, depending on the specifics and if user can see each other's orders or not.

**Comment by jonnybee** [comment]
From the OP: 
> lets say that you have a class BlogPost is there a way to assign read privileges to everybody and write privileges just to the creator of the BlogPost instance?

Given this use case - anyone can Fetch but only creator can edit and save.
